WIP: Add databricks permissions command

areese opened the pull request:
This is a public preview api

Commit:
Add databricks permissions command
This is a public preview api

Commit:
Remove extra file

Commit:
Fix acidental removal of tests/__init__.py

Commit:
tests/permissions/__init__.py was missing the license

gauthamsunjay commented on databricks_cli/permissions/cli.py:
```diff
@@ -0,0 +1,188 @@
+        value = service_name
+    else:
+        click.echo(
+            'Invalid permission-level must be one of {}'.format([e.value for e in PermissionLevel]))
```
`nit`: This is PermissionType and not PermissionLevel.

areese replied:
Ty, fixed.

gauthamsunjay commented on databricks_cli/permissions/cli.py:
```diff
@@ -0,0 +1,188 @@
+        return
+
+    click.echo(pretty_format(perms_api.add_permissions(object_type, object_id, all_permissions)))
+
```
We should add support for the following:
1. Set / update permissions
2. Ability to specify multiple permissions (not sure if this is possible already)

areese replied:
I'll take note of that.

I also need to add some tests, and I'll cover those as part of the tests.

gauthamsunjay commented on tests/permissions/test_cli.py:
```diff
@@ -0,0 +1,84 @@
+    }
+}
+
+
```
Can we add tests for other operations as well?

areese replied:
Yes.  :). Oversight.

I'm trying to collect all of the bits I have, and get them pushed.
Once there are tests and better coverage, we can take another look.

gauthamsunjay reviewed (commented):
Thanks for the PR. Left some comments.

areese commented on databricks_cli/permissions/api.py:
```diff
@@ -0,0 +1,224 @@
+        }
+
+
+class PermissionsApi(object):
```
Thank you, that's a good point I'll fix it.

areese replied:
I'll add update, but looking at https://docs.databricks.com/dev-tools/api/latest/permissions.html#operation/update-all-notebook-permissions

I'm kinda skeptical of set_permissions.  The issue I have is it's dangerous, you could wipe all permission out everywhere if you are not careful.

I'm going to leave that out of this pr, and start a thread with engineering to determine if that's something that should be exposed.

areese commented on databricks_cli/permissions/api.py:
```diff
@@ -0,0 +1,224 @@
+        return self.client.add_permissions(object_type=PermissionTargets[object_type].value,
+                                           object_id=object_id, data=permissions.to_dict())
+
+    def get_id_for_directory(self, path):
```
That's correct, not sure how this ended up there.

areese replied:
This has been moved, good catch, ty.

Commit:
Merge branch 'master' into add-permissions

Commit:
WIP: Add missing bits in the workspace API
Move get_id_for_directory from PermissionsApi to WorkspaceApi

Commit:
Fix an issue where the if statements were checking type but raising an error that said level.  :(

Commit:
Remove WorkspaceApi import

Commit:
Fix a typo in WorkspaceFileInfo.  We cannot change the order of arguments, as that breaks existing code

Commit:
get_id_for_directory was moved from the permissions api to the workspace api

areese commented:
This is still WIP, there are a number of requests from @gauthamsunjay that need to be addressed, once those are fixed, we can review again.

Commit:
Fix lint errors about un-grouped imports

Commit:
Add some documentation showing all of the various urls that the permissions api calls to make life easier

Commit:
databricks permissions:
Move from argument to option so help can be provided.
Share the same help text across all of the databricks permissions functions

Commit:
databricks cli permissions
Move from --object_type to --object-id to be inline with other commands
Fix tests to validate the test output the same way clusters does

Commit:
Fix a lint error in databricks_cli/sdk/permissions.py caused by the type hint not being the first line of the functions
Remove an unused variable from tests/permissions/test_cli.py

Commit:
databricks permissions:
1. Reuse the same help string for all object types
2. Update the group help to display the object types, it could be hard coded or displayed a lot cleaner.
3. object-id is always required, even on list, need to find out why
4. Remove if not object_type checks, as object type is a required option
5. Add mostly complete tests for list-permission-types, I do not have registered-models or instance-pools data available

Commit:
databricks permssions
Fix line too long lint errors.
Fix errors looking up instance-pools in the enum

Commit:
Fix an issue with OneOfOption where the error would say Missing None.
$ databricks permissions add  --object-type clusters --object-id foo
Error: Missing None.  One of ['group-name', 'user-name', 'service-name'] must be provided.

$ databricks permissions add  --object-type clusters --object-id foo
Error: Missing option '--group-name'.  One of ['group-name', 'user-name', 'service-name'] must be provided.

Commit:
databricks permssions
Cleanup ordering of commands added to the group
Remove checks that are done by click and are not required in databricks permissions add

Commit:
databricks permissions
Rename databricks_cli/permissions/api.py:Lookups to PermissionsLookup
PermissionsLookup.from_name should raise a useful error with a message instead of KeyError
Remove more error checking that is done by click.
--permission-level should be checked by click to provide a useful error

Commit:
databricks permissions
Use click.Choice to validate permission levels
Add test for databricks permissions add clusters manage

Commit:
Add more tests of the permissions cli, still need to test the permissions api and service better

Commit:
Fix lint errors

Commit:
Merge remote-tracking branch 'upstream/master' into add-permissions

Commit:
databricks permissions
Allow PermissionsObject to init with constructor
databricks permissions ls should require path, and path should have help
Add more tests

Commit:
databricks permissions:
Remove dead code that wasn't called
Update tests to cover all of the commands

Commit:
Checkpoint trying to cleanup how the permissions are managed to make detecting bad combinations easier

areese committed:
Merge branch 'master' into add-permissions

nfx commented on databricks_cli/permissions/api.py:
```diff
@@ -0,0 +1,343 @@
+
+
+class PermissionTargets(Enum):
+    clusters = 'clusters'
```
logically this should rather be a dictionary or list of tuples, because it's mixing two different things under same enumeration.

e.g. `cluster` is the object type and `clusters` is the prefix for object it,

nfx commented on databricks_cli/permissions/api.py:
```diff
@@ -0,0 +1,343 @@
+        return ', '.join([e.value for e in PermissionLevel])
+
+
+class BasicPermissions(object):
```
it looks over-engineered. please simplify. 

technically, it all can be replaced by single simple lookup structure.

nfx commented on databricks_cli/permissions/cli.py:
```diff
@@ -0,0 +1,178 @@
+@click.command(context_settings=CONTEXT_SETTINGS,
+               short_help='Get permissions for an item.  ' + POSSIBLE_OBJECT_TYPES)
+@click.option('--object-type', required=True, help=POSSIBLE_OBJECT_TYPES)
+@click.option('--object-id', required=True, help='object id to require permission about')
```
why not providing `one of` kind of arguments? e.g. `--cluster-id`, `--job-id`, `--tokens`, `--passwords`, `--cluster-policy-id` and etc? CLI is supposed to simplify usage for admins and not always be 1-1 mapping to API resources.

nfx commented on databricks_cli/permissions/cli.py:
```diff
@@ -0,0 +1,178 @@
+@click.option('--group-name', metavar='<string>', cls=OneOfOption,
+              one_of=GROUP_USER_SERVICE_OPTIONS)
+@click.option('--user-name', metavar='<string>', cls=OneOfOption, one_of=GROUP_USER_SERVICE_OPTIONS)
+@click.option('--service-name', metavar='<string>', cls=OneOfOption,
```
it's not really a service name, but `--service-principal`

nfx commented on databricks_cli/permissions/exceptions.py:
```diff
@@ -0,0 +1,26 @@
+# limitations under the License.
+
+
+class PermissionsError(Exception):
```
why do we need a file for two lines of code? why it cannot be in the same file as where it's thrown from?

nfx commented on databricks_cli/sdk/permissions_service.py:
```diff
@@ -0,0 +1,95 @@
+from typing import Optional
+
+from databricks_cli.sdk.preview_service import PreviewService
```
there's no such a thing as `preview` service. please remove or rename to something like `PreviewStability`

areese replied:
That's reasonable.

I'm probably going  to use the newer openapi spec that is part of the rest specification and rewrite this differently.

nfx reviewed (changes requested):
please see comments

areese commented:
FYI: This is stale, it has semi-useful code but has a lot to be addressed.
I'm going to leave it open, but I'm not sure how fast I'll come back to it.

chrisst commented:
This repository has been archived and is no longer maintained. The legacy `databricks/databricks-cli` has been fully replaced by the new Databricks CLI at https://github.com/databricks/cli.
